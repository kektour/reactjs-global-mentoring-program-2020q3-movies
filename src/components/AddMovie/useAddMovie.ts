import { AxiosError } from 'axios';
import { useCallback, useState } from 'react';
import { Movie } from '../../models/movie';
import { createMovie as cM } from '../../services/movies';

type UseAddMovie = {
  error: Record<string, string> | null;
  clearError: () => void;
  createMovie: (m: Omit<Movie, 'id'>) => Promise<void>;
};

interface PostMovieErrorResponse {
  messages: Array<string>;
}

export const useAddMovie = (): UseAddMovie => {
  const [error, setError] = useState<Record<string, string> | null>(null);
  const createMovie = useCallback(async (m: Omit<Movie, 'id'>) => {
    try {
      await cM(m);
    } catch (err) {
      const typedErr = err as AxiosError<PostMovieErrorResponse>;
      const errorMessages = getErrorMessages(typedErr.response!.data.messages);
      setError(errorMessages);
      throw new Error('Create movie error');
    }
  }, []);
  const clearError = useCallback(() => setError(null), []);

  return {
    error,
    clearError,
    createMovie,
  };
};

const getErrorMessages = (arr: Array<string>): Record<string, string> =>
  arr.reduce((obj, v) => {
    let [fieldName] = v.split(' ');
    fieldName = fieldName.replace(/"/g, '');
    return {
      ...obj,
      [fieldName]: v,
    };
  }, {});
