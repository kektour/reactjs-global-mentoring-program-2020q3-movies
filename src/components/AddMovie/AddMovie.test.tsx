import React from 'react';
import ReactDOM from 'react-dom';
import { render, screen, wait } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { AddMovie } from './AddMovie';
import { act } from 'react-dom/test-utils';
import * as hooks from './useAddMovie';

describe('AddMovie', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((e) => e) as any;

    const body = document.createElement('body');
    const div = document.createElement('div');
    const spyQuerySelector = jest.fn((selector: string) => {
      if (selector === '#root') {
        return div;
      } else if (selector === 'body') {
        return body;
      }
    });
    Object.defineProperty(global.document, 'querySelector', { value: spyQuerySelector });
  });

  it('Should render component correctly', () => {
    const { asFragment } = render(<AddMovie open={true} onClose={() => null} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should interact with the form', async () => {
    const { asFragment, container } = render(<AddMovie open={true} onClose={() => null} />);
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Title here'), 'Title Foo'));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Release Date here'), '2019-09-04'));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Movie Preview here'), 'Movie Preview Foo'));
    act(() => userEvent.selectOptions(container.querySelector('select[name=genres]') as Element, ['baz']));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Overview here'), 'Overview Foo'));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Runtime here'), '123'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should call API on submit form', async () => {
    const createMovieMock = jest.fn().mockImplementation(() => Promise.resolve());
    const clearErrorMock = jest.fn();
    const handleCloseMock = jest.fn();
    const useAddMovieMock = jest.spyOn(hooks, 'useAddMovie').mockImplementation(() => ({ error: null, clearError: clearErrorMock, createMovie: createMovieMock }));

    const { container } = render(<AddMovie open={true} onClose={handleCloseMock} />);
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Title here'), 'Title Foo'));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Release Date here'), '2019-09-04'));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Movie Preview here'), 'Movie Preview Foo'));
    act(() => userEvent.selectOptions(container.querySelector('select[name=genres]') as Element, ['baz']));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Overview here'), 'Overview Foo'));
    await act(async () => await userEvent.type(screen.getByPlaceholderText('Runtime here'), '123'));
    act(() => userEvent.click(container.querySelector('button[type=submit]') as Element));

    await wait(() => {
      expect(createMovieMock).toHaveBeenCalledWith({
        tagline: 'tagline',
        vote_average: 0,
        vote_count: 0,
        budget: 0,
        revenue: 0,
        title: 'Title Foo',
        release_date: '2019-09-04',
        poster_path: 'Movie Preview Foo',
        genres: ['baz'],
        overview: 'Overview Foo',
        runtime: '123',
      });
    });
    expect(handleCloseMock).toHaveBeenCalled();
    expect(clearErrorMock).toHaveBeenCalled();

    useAddMovieMock.mockRestore();
  });

  it('Should show errors', async () => {
    const useAddMovieMock = jest.spyOn(hooks, 'useAddMovie').mockImplementation(
      () =>
        ({
          error: {
            title: '"title" error title field',
            genres: '"genres" error genres field',
            overview: '"overview" error overview field',
          },
        } as any)
    );

    const { asFragment } = render(<AddMovie open={true} onClose={() => null} />);
    expect(asFragment()).toMatchSnapshot();

    useAddMovieMock.mockRestore();
  });
});
