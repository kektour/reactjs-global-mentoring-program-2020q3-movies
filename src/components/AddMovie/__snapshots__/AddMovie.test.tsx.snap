// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovie Should interact with the form 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Add Movie
        </h2>
        <form
          class="form"
        >
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Title
            </p>
            <input
              class="input"
              name="title"
              placeholder="Title here"
              value="Title Foo"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Release Date
            </p>
            <input
              class="input"
              name="release_date"
              placeholder="Release Date here"
              type="date"
              value="2019-09-04"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Preview
            </p>
            <input
              class="input"
              name="poster_path"
              placeholder="Movie Preview here"
              value="Movie Preview Foo"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Gende
            </p>
            <select
              class="root selectRoot selectInput"
              multiple=""
              name="genres"
            >
              <option
                disabled=""
                hidden=""
                value=""
              >
                Choose here
              </option>
              <option
                class="option selectOption"
                value="foo"
              >
                Foo
              </option>
              <option
                class="option selectOption"
                value="bar"
              >
                Bar
              </option>
              <option
                class="option selectOption"
                value="baz"
              >
                Baz
              </option>
            </select>
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Overview
            </p>
            <input
              class="input"
              name="overview"
              placeholder="Overview here"
              value="Overview Foo"
            />
          </div>
          <div
            class="root inputRoot lastInputRoot"
          >
            <p
              class="title"
            >
              Runtime
            </p>
            <input
              class="input"
              name="runtime"
              placeholder="Runtime here"
              value="123"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="buttonReset"
              type="reset"
            >
              Reset
            </button>
            <button
              class="buttonSave"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddMovie Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Add Movie
        </h2>
        <form
          class="form"
        >
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Title
            </p>
            <input
              class="input"
              name="title"
              placeholder="Title here"
              value=""
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Release Date
            </p>
            <input
              class="input"
              name="release_date"
              placeholder="Release Date here"
              type="date"
              value="2020-10-24"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Preview
            </p>
            <input
              class="input"
              name="poster_path"
              placeholder="Movie Preview here"
              value="https://m.media-amazon.com/images/M/MV5BYzg0NGM2NjAtNmIxOC00MDJmLTg5ZmYtYzM0MTE4NWE2NzlhXkEyXkFqcGdeQXVyMTA4NjE0NjEy._V1_.jpg"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Gende
            </p>
            <select
              class="root selectRoot selectInput"
              multiple=""
              name="genres"
            >
              <option
                disabled=""
                hidden=""
                value=""
              >
                Choose here
              </option>
              <option
                class="option selectOption"
                value="foo"
              >
                Foo
              </option>
              <option
                class="option selectOption"
                value="bar"
              >
                Bar
              </option>
              <option
                class="option selectOption"
                value="baz"
              >
                Baz
              </option>
            </select>
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Overview
            </p>
            <input
              class="input"
              name="overview"
              placeholder="Overview here"
              value=""
            />
          </div>
          <div
            class="root inputRoot lastInputRoot"
          >
            <p
              class="title"
            >
              Runtime
            </p>
            <input
              class="input"
              name="runtime"
              placeholder="Runtime here"
              value="0"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="buttonReset"
              type="reset"
            >
              Reset
            </button>
            <button
              class="buttonSave"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddMovie Should show errors 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Add Movie
        </h2>
        <form
          class="form"
        >
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Title
            </p>
            <input
              class="input"
              name="title"
              placeholder="Title here"
              value=""
            />
          </div>
          <p
            style="color: red;"
          >
            "title" error title field
          </p>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Release Date
            </p>
            <input
              class="input"
              name="release_date"
              placeholder="Release Date here"
              type="date"
              value="2020-10-24"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Preview
            </p>
            <input
              class="input"
              name="poster_path"
              placeholder="Movie Preview here"
              value="https://m.media-amazon.com/images/M/MV5BYzg0NGM2NjAtNmIxOC00MDJmLTg5ZmYtYzM0MTE4NWE2NzlhXkEyXkFqcGdeQXVyMTA4NjE0NjEy._V1_.jpg"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Gende
            </p>
            <select
              class="root selectRoot selectInput"
              multiple=""
              name="genres"
            >
              <option
                disabled=""
                hidden=""
                value=""
              >
                Choose here
              </option>
              <option
                class="option selectOption"
                value="foo"
              >
                Foo
              </option>
              <option
                class="option selectOption"
                value="bar"
              >
                Bar
              </option>
              <option
                class="option selectOption"
                value="baz"
              >
                Baz
              </option>
            </select>
          </div>
          <p
            style="color: red;"
          >
            "genres" error genres field
          </p>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Overview
            </p>
            <input
              class="input"
              name="overview"
              placeholder="Overview here"
              value=""
            />
          </div>
          <p
            style="color: red;"
          >
            "overview" error overview field
          </p>
          <div
            class="root inputRoot lastInputRoot"
          >
            <p
              class="title"
            >
              Runtime
            </p>
            <input
              class="input"
              name="runtime"
              placeholder="Runtime here"
              value="0"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="buttonReset"
              type="reset"
            >
              Reset
            </button>
            <button
              class="buttonSave"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;
